include/bout/petsc_interface.hxx

```diff
@@ -42,64 +42,223 @@
+
+  const Region<ind_type>& getRegionAll() const { return regionAll; }
+  const Region<ind_type>& getRegionNobndry() const { return indices.getRegion("RGN_NOBNDRY"); }
+  const Region<ind_type>& getRegionBndry() const { return regionBndry; }
+  const Region<ind_type>& getRegionLowerY() const { return regionLowerY; }
+  const Region<ind_type>& getRegionUpperY() const { return regionUpperY; }
+  const Region<ind_type>& getRegionInnerX() const { return regionInnerX; }
+  const Region<ind_type>& getRegionOuterX() const { return regionOuterX; }
+
+  bool sparsityPatternAvailable() const {
+    return stencils.getNumParts() > 0;
+  }
+  
+  std::vector<int> getNumDiagonal() {
+    ASSERT2(sparsityPatternAvailable());
+    if (!sparsityCalculated) {
+      calculateSparsity();
+    }
+    return numDiagonal;
+  }
+
+  std::vector<int> getNumOffDiagonal() {
+    ASSERT2(sparsityPatternAvailable());
+    if (!sparsityCalculated) {
+      calculateSparsity();
+    }
+    return numOffDiagonal;
+  }
 
-  // Mark the globalIndexer instance to be recreated next time it is
-  // requested (useful for unit tests)
-  static void recreateGlobalInstance();
+  int size() const {
+    return regionAll.size();
+  }
 
 protected:
-  GlobalIndexer(Mesh* localmesh);
-  Field3D& getIndices3D() { return indices3D; }
-  Field2D& getIndices2D() { return indices2D; }
-  FieldPerp& getIndicesPerp() { return indicesPerp; }
+  T& getIndices() { return indices; }
 
 private:
+  void insertIndex(const ind_type i, std::set<ind_type> &allInds,
+		   std::set<ind_type> &newInds) {
+    auto result = allInds.insert(i);
+    if (result.second) {
+      newInds.insert(i);
+    }
+  }
+
   /// This gets called by initialiseTest and is used to register
   /// fields with fake parallel meshes.
-  virtual void registerFieldForTest(FieldData& f);
-  virtual void registerFieldForTest(FieldPerp& f);
+  virtual void registerFieldForTest(T& UNUSED(f)) {
+    // This is a place-holder which does nothing. It can be overridden
+    // by descendent classes if necessary to set up testing.
+    return;
+  }
+  
+  void calculateSparsity() {
+    numDiagonal = std::vector<int>(size());
+    numOffDiagonal = std::vector<int>(size(), 0);
+
+    // Set initial guess for number of on-diagonal elements
+    BOUT_FOR(i, regionAll) {
+    	numDiagonal[getGlobal(i) - globalStart] = stencils.getStencilSize(i);
+    }
+
+    BOUT_FOR_SERIAL(i, regionBndry) {
+      if (!isLocal(i)) {
+	for (const auto &j: stencils.getIndicesWithStencilIncluding(i)) {
+          if (isLocal(j)) {
+            const int n = getGlobal(j) - globalStart;
+            numDiagonal[n] -= 1;
+            numOffDiagonal[n] += 1;
+          }
+        }
+      }
+    }
+  }
 
   PetscLib lib;
   Mesh* fieldmesh;
 
   /// Fields containing the indices for each element (as reals)
-  Field3D indices3D;
-  Field2D indices2D;
-  FieldPerp indicesPerp;
+  T indices;
+  PetscInt globalStart, globalEnd;
```

Comment:
Am I right in saying `globalEnd` is the inclusive end? Best to leave a comment if so

Reply:
Yes, that's right.